resteasy-core/src/main/java/org/jboss/resteasy/specimpl/ResteasyUriBuilderImpl.java

```diff
@@ -1088,4 +1087,86 @@ public UriBuilder resolveTemplatesFromEncoded(Map<String, Object> templateValues
             throw new IllegalArgumentException(Messages.MESSAGES.mapKeyNull());
         return uriTemplate(buildCharSequence(templateValues, true, true, true));
     }
+
+    /**
+     * The URI spec rfc3986, section 3.4 Query notes it may not be possible,
+     * "to distinguish query data from path data when looking for hierarchical separators."
+     * (e.g. http://http://127.0.0.1:8080/something/{string:[a-z]?}/cust?q={"status":"GOLD"})
+     * Parsing the uri and differentiating the regex expression "{string:[a-z]?}" from
+     * the query expression "cust?q={"status":"GOLD"}" is difficult. This method performs
+     * extra processing to identify path regex expressions from query expressions and
+     * encode them as appropriate.
+     *
+     * @param group5 uri path after "http://host:port/
+     * @param group6 proposed query text from "?" designator and following text
+     * @param group7 proposed query text following "?"
+     */
+    private void processRegexVsQueryText(String group5, String group6, String group7) {
+
+        String tmpGroup5 = group5;
+        String tmpGroup6 = group6;
+
+        if (group5 != null && group6 != null) {
+            int posBracketClose = group6.indexOf("}");
+            if (posBracketClose > -1) {
+                int posSlash = group5.lastIndexOf("/");
+                if (group5.regionMatches(posSlash + 1, "{", 0, 1)) {
+                    tmpGroup5 = group5 + group6.substring(0, posBracketClose + 1);
+                    String tmpTmpGroup6 = group6.substring(posBracketClose + 1);
+
+                    int posQmark = tmpTmpGroup6.indexOf("?");
+                    if (posQmark != -1) {
+                        tmpGroup6 = tmpTmpGroup6.substring(posQmark + 1);
+                    } else {
+                        tmpGroup6 = "";
+                    }
```

Comment:
This breaks the `UriBuilderTest`. The pattern `http://{host}/x/y/{path}?{q}={qval}` is not parsed correctly. The `tmpGroup6` ends up being set to "" leaving off the `?` and `={qval}`. 

It might also be helpful to name the parameters something other than "groupX" so it's more self documenting.